Token auth

ekzema opened the pull request:
https://app.asana.com/0/1159181417647996/1159181417648022

maksym-bobmm committed:
Add gems devise, rspec-rails

Some configs

maksym-bobmm committed:
Add device-jwt gem

Some configs

ekzema committed:
Add error and success response

ekzema committed:
Add action create in users controller

ekzema committed:
Add gem omniauth-facebook

ekzema committed:
Add gem koala and validate login

ekzema committed:
Add Graph Api

ekzema committed:
Add action Sessions#create and generate api token

ekzema committed:
Add action check authorize

drkmen commented on Gemfile:
```diff
@@ -3,8 +3,17 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
+
+gem 'koala', '~> 3.0'
+
+gem 'dotenv-rails', groups: [:development, :test]
```
there are already created group:
```ruby 
group :development, :test do
```
move it there

drkmen commented on app/controllers/application_controller.rb:
```diff
@@ -1,7 +1,39 @@
     render json: json, code: 200
   end
+
+  def render_success_response(data = nil, msg = 'Ok')
```
Do not abbreviate args and variable name until it's really needed.
Success response doesn't require message at all. It returns data instead

drkmen commented on app/controllers/application_controller.rb:
```diff
@@ -1,7 +1,39 @@
+    }, status: :ok
+  end
+
+  def render_error_response(msg = 'Bad Request', status = :bad_request)
```
Do not abbreviate args and variable name until it's really needed.

drkmen commented on app/controllers/application_controller.rb:
```diff
@@ -1,7 +1,39 @@
+  end
+
+  def configure_permitted_parameters
+    added_attrs = [:phone, :email, :password, :password_confirmation]
```
```ruby
%i[phone email password password_confirmation]
```

drkmen commented on app/controllers/application_controller.rb:
```diff
@@ -1,7 +1,39 @@
+
+  def check_authorize
+    token = request.headers['token']
+    return render_error_response unless token
```
add empty line after guard clause

drkmen commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,11 @@
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  skip_before_action :check_authorize, only: [:facebook]
```
use `%i[]`

drkmen commented on app/controllers/users/registrations_controller.rb:
```diff
@@ -0,0 +1,11 @@
+
+  # POST api/users/signup
+  def create
+    result = Users::RegistrationService.new(params).perform
```
```ruby
Users::RegistrationService.perform(params)
```

drkmen commented on app/controllers/users/sessions_controller.rb:
```diff
@@ -0,0 +1,11 @@
+
+  # POST api/users/login
+  def create
+    result = Users::SessionService.new(warden.authenticate).perform
```
```ruby
Users::SessionService.perform(warden.authenticate)
```

drkmen commented on app/lib/json_web_token.rb:
```diff
@@ -0,0 +1,13 @@
+class JsonWebToken
+  SECRET_KEY = Rails.application.secrets.secret_key_base.to_s
```
always `.freeze` constants

drkmen commented on app/lib/json_web_token.rb:
```diff
@@ -0,0 +1,13 @@
+class JsonWebToken
+  SECRET_KEY = Rails.application.secrets.secret_key_base.to_s
+
+  def self.encode(payload, exp = 24.hours.from_now)
```
do not abbreviate args and variable names until it's really needed

drkmen commented on app/models/user.rb:
```diff
@@ -0,0 +1,43 @@
+                      message: 'Phone numbers must be in xxx-xxx-xxxx format.', if: :phone_present?
+  validate :email_or_phone
+  validates :password, confirmation: true
+  # validates :phone, presence: true
```
do not keep commented code

drkmen commented on app/models/user.rb:
```diff
@@ -0,0 +1,43 @@
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  include Devise::JWT::RevocationStrategies::JTIMatcher
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :jwt_authenticatable, jwt_revocation_strategy: Devise::JWT::RevocationStrategies::Null
```
this line seems to long, please use rubocop

drkmen commented on app/models/user.rb:
```diff
@@ -0,0 +1,43 @@
+  def self.find_for_database_authentication(warden_conditions)
+    conditions = warden_conditions.dup
+    if login = conditions.delete(:login)
+      where(conditions.to_h).where(["lower(phone) = :value OR lower(email) = :value", { :value => login.downcase }]).first
```
this line seems to long, please use rubocop

drkmen commented on app/models/user.rb:
```diff
@@ -0,0 +1,43 @@
+  end
+
+  def email_present?
+    !email.empty?
```
What happens if email is `nil`?
```ruby
email.present?
```

drkmen commented on app/models/user.rb:
```diff
@@ -0,0 +1,43 @@
+  end
+
+  def phone_present?
+    !phone.empty?
```
```ruby
phone.present?
```

drkmen commented on app/services/application_service.rb:
```diff
@@ -9,5 +9,12 @@ def self.perform(params = {}, &block)
+
+  def jwt_decode(token)
+    JsonWebToken.decode(token)
+  end
```
This is not a place for those two methods
Create new service `Users::ApllicationService` and place them there. Inherit it from `ApplicationService`
and all services which is need those methods will inherits from `Users::ApllicationService`

drkmen commented on app/services/users/omniauth_facebook_service.rb:
```diff
@@ -0,0 +1,35 @@
+module Users
+  class OmniauthFacebookService < ApplicationService
+    def initialize(token)
```
already defined in `ApplicationService`

drkmen commented on app/services/users/omniauth_facebook_service.rb:
```diff
@@ -0,0 +1,35 @@
+module Users
+  class OmniauthFacebookService < ApplicationService
```
Add service documentation. Check example service

drkmen commented on app/services/users/omniauth_facebook_service.rb:
```diff
@@ -0,0 +1,35 @@
+      @token = token
+    end
+
+    def perform
```
rename to `call` (i've update service example)

drkmen commented on app/services/users/omniauth_facebook_service.rb:
```diff
@@ -0,0 +1,35 @@
+        message = e.message.instance_of?(String) ? {error: e.message} : e.message
+        OpenStruct.new(success?: false, errors: message)
+      end
+    end
```
```ruby 
def validate_facebook_token
     graph = Koala::Facebook::API.new(@token)
     get_user(graph.get_object("me?fields=email"))
rescue => e
    message = e.message.instance_of?(String) ? {error: e.message} : e.message
    OpenStruct.new(success?: false, errors: message)
end
```

> message = e.message.instance_of?(String) ? {error: e.message} : e.message

Why do you need this? `e.message` always should be a `String`

drkmen commented on app/services/users/registration_service.rb:
```diff
@@ -0,0 +1,13 @@
+  class RegistrationService < ApplicationService
+    def initialize(params)
+      @params = params.require(:user).permit(:email, :phone, :password, :password_confirmation)
+    end
```
Already defined in `ApplicationService`
`params` should be permitted in controller

drkmen commented on app/services/users/session_service.rb:
```diff
@@ -0,0 +1,12 @@
+module Users
+  class SessionService < ApplicationService
+    def initialize(user)
```
Already defined in `ApplicationService`

drkmen commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
```
why it's empty?

drkmen commented:
@faceitruby/developers 
1. Use rubocop
2. All code should be covered with tests

ekzema committed:
fix %i[]

ekzema committed:
Merge branch 'development' into token_auth

ekzema committed:
delete comment

maksym-bobmm committed:
Change email and phone present methods

Reduce few lines length

maksym-bobmm committed:
Change variable to full name

Add freeze to constant

maksym-bobmm committed:
Fix pull request

Add service Users::ApllicationService
Change perform to call in services
Some refactoring with rubocop
Remove initialize methods which are exists in parent class
Change service call to perform method
Add service descriptions
Add some tests
Add shoulda-matchers gem
Add migration for fix creating few users
Some refactoring with rubocop
Remove unneeded field(password_confirmation)
